src/native/minipal/thread.c

```diff
@@ -0,0 +1,130 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#define _GNU_SOURCE
+
+#include <stdbool.h>
+#include "utils.h"
+#include "thread.h"
+
+#ifdef TARGET_UNIX
+
+// Async safe lock free thread map for use in signal handlers
+
+struct ThreadEntry
+{
+    size_t osThread;
+    void* pThread;
+};
+
+#define MAX_THREADS_IN_SEGMENT 256
+
+struct ThreadSegment
+{
+    struct ThreadEntry entries[MAX_THREADS_IN_SEGMENT];
+    struct ThreadSegment* pNext;
+};
+
+static struct ThreadSegment *s_pAsyncSafeThreadMapHead = NULL;
+
+bool minipal_insert_thread_into_async_safe_map(size_t osThread, void* pThread)
+{
+    size_t startIndex = osThread % MAX_THREADS_IN_SEGMENT;
+
+    struct ThreadSegment* pSegment = s_pAsyncSafeThreadMapHead;
+    struct ThreadSegment** ppSegment = &s_pAsyncSafeThreadMapHead;
+    while (true)
+    {
+        if (pSegment == NULL)
+        {
+            // Need to add a new segment
+            struct ThreadSegment* pNewSegment = (struct ThreadSegment*)malloc(sizeof(struct ThreadSegment));
+            if (pNewSegment == NULL)
+            {
+                // Memory allocation failed
+                return false;
+            }
+
+            memset(pNewSegment, 0, sizeof(struct ThreadSegment));
+            struct ThreadSegment* pExpected = NULL;           
+            if (!__atomic_compare_exchange_n(
+                ppSegment,
+                &pExpected,
+                pNewSegment,
+                false /* weak */,
+                __ATOMIC_RELEASE  /* success_memorder */,
+                __ATOMIC_RELAXED /* failure_memorder */))
+            {
+                // Another thread added the segment first
+                free(pNewSegment);
+                pNewSegment = *ppSegment;
```

Comment:
Instead of reaching back up and accessing a potentially changing value, this API writes the non-equal failure value back into `pExpected`. I think the more correct operation would be `pNewSegment = *pExpected;`